tests/unit/services/test_request.py

```diff
@@ -0,0 +1,109 @@
+#  under the terms of the GNU Lesser General Public License version 3, as
+#  published by the Free Software Foundation.
+#
+#  This program is distributed in the hope that it will be useful, but WITHOUT
+#  ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+#  SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.
+#  See the GNU Lesser General Public License for more details.
+#
+#  You should have received a copy of the GNU Lesser General Public License
+#  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Unit tests for the Request service."""
+from unittest.mock import call
+
+import craft_cli.pytest_plugin
+import pytest
+import pytest_check
+import responses
+from hypothesis import HealthCheck, given, settings, strategies
+
+
+@responses.activate
+@settings(suppress_health_check=[HealthCheck.function_scoped_fixture])
+@given(data=strategies.binary())
+def test_download_chunks(tmp_path, request_service, data):
+    responses.add(
+        responses.GET,
+        "http://example/file",
+        body=data,
+        headers={"Content-Length": str(len(data))},
+    )
+    output_file = tmp_path / "file"
+
+    downloader = request_service.download_chunks("http://example/file", output_file)
+    size = next(downloader)
+    dl_size = sum(downloader)
+
+    pytest_check.equal(int(size), len(data), "Downloaded size is incorrect")
+    pytest_check.equal(dl_size, len(data), "Downloaded size is incorrect")
+    pytest_check.equal(output_file.read_bytes(), data, "Download data is incorrect")
+
+    downloader = request_service.download_chunks("http://example/file", tmp_path)
+    size = next(downloader)
+    dl_size = sum(downloader)
+
+    pytest_check.equal(int(size), len(data), "Downloaded size is incorrect")
+    pytest_check.equal(dl_size, len(data), "Downloaded size is incorrect")
+    pytest_check.equal(output_file.read_bytes(), data, "Download data is incorrect")
+
+
+@responses.activate
+@settings(suppress_health_check=[HealthCheck.function_scoped_fixture])
+@given(data=strategies.binary(min_size=1))
+def test_download_with_progress(
+    tmp_path, emitter: craft_cli.pytest_plugin.RecordingEmitter, request_service, data
+):
+    emitter.interactions = []  # Reset the emitter so we can re-run with hypothesis.
+    responses.add(
+        responses.GET,
+        "http://example/file",
+        body=data,
+        headers={"Content-Length": str(len(data))},
+    )
+    output_file = tmp_path / "file"
+
+    request_service.download_with_progress("http://example/file", tmp_path)
+
+    pytest_check.equal(output_file.read_bytes(), data, "Download data is incorrect")
+    emitter.assert_interactions(
+        [
+            call("progress_bar", "Downloading http://example/file", len(data)),
+            call("advance", len(data)),
+        ]
+    )
+
+
+@responses.activate
+@pytest.mark.parametrize(
+    "downloads",
+    [
+        {
+            "http://example/file": b"abc",
```

Comment:
What about testing an empty file?

Reply:
Done! But I've only tested it here because it requires more logic in the test (empty files don't get an `advance` call).